malformed packet received on agent peer-port crashes agent

**Describe the bug**
The bug happens when random packet received on the peer-port of the kraken-agent, instead of a p2p message from other peers.
When receiving a packet from the `peer-port` , this code will retrieve the content and validate the msg:
https://github.com/uber/kraken/blob/master/lib/torrent/scheduler/conn/message.go#L126-L144
However, if some malformed packet accidentally passes those 4 checks, and finally constructs an empty p2p.Message struct, the empty msg struct will then go to here:
https://github.com/uber/kraken/blob/d8d5afbe4dda5cacdd1c5de896ff1e0409910992/lib/torrent/scheduler/conn/handshaker.go#L97
m.Type checking is fine, since the default value for int32 is 0 - unexpectedly matches this check; but when retrieving m.Bitfield, there would be a nil pointer panic.

**To Reproduce**
Try to send an empty p2p msg mocking the peer msg


**Expected behavior**
kraken-agent will panic with `nil pointer`.

**Environments**
docker container running on bare metal.

**Additional context**
panic strack trace:
```
panic: runtime error: invalid memory address or nil pointer dereference
signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x8148f8

goroutine 200 [running]:
github.com/uber/kraken/lib/torrent/scheduler/conn.handshakeFromP2PMessage(0xc0004024b0, 0xc00000e1c0, 0x12a05f200, 0xc0004024b0)
	/go/src/github.com/uber/kraken/lib/torrent/scheduler/conn/handshaker.go:101 +0x48
github.com/uber/kraken/lib/torrent/scheduler/conn.(*Handshaker).readHandshake(0xc000404280, 0xccbda0, 0xc00000e1c0, 0xc000200c40, 0xc000200cb0, 0xc000200d20)
	/go/src/github.com/uber/kraken/lib/torrent/scheduler/conn/handshaker.go:302 +0xe9
github.com/uber/kraken/lib/torrent/scheduler/conn.(*Handshaker).Accept(0xc000404280, 0xccbda0, 0xc00000e1c0, 0xc000201180, 0xc0002011f0, 0xc000201260)
	/go/src/github.com/uber/kraken/lib/torrent/scheduler/conn/handshaker.go:228 +0x43
github.com/uber/kraken/lib/torrent/scheduler.(*scheduler).listenLoop.func1(0xc0000e2500, 0xccbda0, 0xc00000e1c0)
	/go/src/github.com/uber/kraken/lib/torrent/scheduler/scheduler.go:326 +0x4a
created by github.com/uber/kraken/lib/torrent/scheduler.(*scheduler).listenLoop
	/go/src/github.com/uber/kraken/lib/torrent/scheduler/scheduler.go:325 +0x174
```

--- a/lib/torrent/scheduler/conn/handshaker.go
+++ b/lib/torrent/scheduler/conn/handshaker.go
@@ -98,24 +98,28 @@ func handshakeFromP2PMessage(m *p2p.Message) (*handshake, error) {
 	if m.Type != p2p.Message_BITFIELD {
 		return nil, fmt.Errorf("expected bitfield message, got %s", m.Type)
 	}
-	peerID, err := core.NewPeerID(m.Bitfield.PeerID)
+	bitfieldMsg := m.GetBitfield()
+	if bitfieldMsg == nil {
+		return nil, fmt.Errorf("empty bit field")
+	}
+	peerID, err := core.NewPeerID(bitfieldMsg.PeerID)
 	if err != nil {
 		return nil, fmt.Errorf("peer id: %s", err)
 	}
-	ih, err := core.NewInfoHashFromHex(m.Bitfield.InfoHash)
+	ih, err := core.NewInfoHashFromHex(bitfieldMsg.InfoHash)
 	if err != nil {
 		return nil, fmt.Errorf("info hash: %s", err)
 	}
-	d, err := core.NewSHA256DigestFromHex(m.Bitfield.Name)
+	d, err := core.NewSHA256DigestFromHex(bitfieldMsg.Name)
 	if err != nil {
 		return nil, fmt.Errorf("name: %s", err)
 	}
 	bitfield := bitset.New(0)
-	if err := bitfield.UnmarshalBinary(m.Bitfield.BitfieldBytes); err != nil {
+	if err := bitfield.UnmarshalBinary(bitfieldMsg.BitfieldBytes); err != nil {
 		return nil, err
 	}
 	remoteBitfields := make(RemoteBitfields)
-	if err := remoteBitfields.unmarshalBinary(m.Bitfield.RemoteBitfieldBytes); err != nil {
+	if err := remoteBitfields.unmarshalBinary(bitfieldMsg.RemoteBitfieldBytes); err != nil {
 		return nil, err
 	}
 
@@ -124,7 +128,7 @@ func handshakeFromP2PMessage(m *p2p.Message) (*handshake, error) {
 		infoHash:        ih,
 		bitfield:        bitfield,
 		digest:          d,
-		namespace:       m.Bitfield.Namespace,
+		namespace:       bitfieldMsg.Namespace,
 		remoteBitfields: remoteBitfields,
 	}, nil
 }

--- a/lib/torrent/scheduler/conn/handshaker_test.go
+++ b/lib/torrent/scheduler/conn/handshaker_test.go
@@ -30,14 +30,13 @@ import (
 func TestHandshakerSetsConnFieldsProperly(t *testing.T) {
 	require := require.New(t)
 
-	config := ConfigFixture()
-
-	h1 := HandshakerFixture(config)
 	l1, err := net.Listen("tcp", "localhost:0")
 	require.NoError(err)
 	defer l1.Close()
 
+	config := ConfigFixture()
 	namespace := core.TagFixture()
+	h1 := HandshakerFixture(config)
 	h2 := HandshakerFixture(config)
 
 	info := storage.TorrentInfoFixture(4, 1)
@@ -92,7 +91,37 @@ func TestHandshakerSetsConnFieldsProperly(t *testing.T) {
 func TestHandshakerHandlesEmptyBitfield(t *testing.T) {
 	require := require.New(t)
 
-	var msg p2p.Message
-	_, err := handshakeFromP2PMessage(&msg)
-	require.Error(err)
+	l1, err := net.Listen("tcp", "localhost:0")
+	require.NoError(err)
+	defer l1.Close()
+
+	config := ConfigFixture()
+	h1 := HandshakerFixture(config)
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		nc, err := l1.Accept()
+		require.NoError(err)
+
+		_, err = h1.Accept(nc)
+		require.Error(err)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		var msg p2p.Message
+
+		nc, err := net.DialTimeout("tcp", l1.Addr().String(), config.HandshakeTimeout)
+		require.NoError(err)
+		err = sendMessage(nc, &msg)
+		require.NoError(err)
+	}()
+
+	wg.Wait()
 }

--- a/lib/torrent/scheduler/conn/handshaker_test.go
+++ b/lib/torrent/scheduler/conn/handshaker_test.go
@@ -22,6 +22,7 @@ import (
 	"github.com/stretchr/testify/require"
 
 	"github.com/uber/kraken/core"
+	"github.com/uber/kraken/gen/go/proto/p2p"
 	"github.com/uber/kraken/lib/torrent/storage"
 	"github.com/uber/kraken/utils/bitsetutil"
 )
@@ -86,4 +87,12 @@ func TestHandshakerSetsConnFieldsProperly(t *testing.T) {
 	}()
 
 	wg.Wait()
+}
+
+func TestHandshakerHandlesEmptyBitfield(t *testing.T) {
+	require := require.New(t)
+
+	var msg p2p.Message
+	_, err := handshakeFromP2PMessage(&msg)
+	require.Error(err)
 }

